### DISCUSSION TITLE
Low frames using Blackfly S USB3 (522fps) with 1.29.0.5 Spinnaker

### DISCUSSION THREAD

**Discussion by AshhadQazi** [bonsai-rx/bonsai#1845] [Q&A]
Hello!

We have multiple high-fps FLIR cameras that we have been testing with Bonsai. Using the older 1.29.0.5 Spinnaker software, connecting and testing all the cameras are fine. We have a 227 fps Blackfly S as well, and some Grasshopper, but all the frames are capped to 200. Testing the same cameras with the newer 4.0 Spinnaker software, which is incompatible with Bonsai usage, gives us our promised FPS. I have come to the conclusion that the older drivers are not efficient enough to fully optimize the newer cameras/higher fps. Even when deciding to go with 200 fps, when running 10-min tests with 3 cameras attached, the outputted video files vary in length. Their quality remains the same, but some video is either sped up or loss. 

My guess with all of this is just that the old drivers can't handle the high frames, and that Bonsai doesn't seem to know how to processes such high volumes of frames. Recording through Spinview has the same issues, the recording fps stays at 200, but the processing fps tends to fluctuate around 160-180 fps. I did some digging, and found when optimizing Bonsai recording to use the GPU, it has yielded better results. However, how does one accomplish that?

This is how I have been recording:
![image](https://github.com/bonsai-rx/bonsai/assets/170653199/aeadea2f-931b-4e37-a7a7-53284ed63adb)

Whilst running tests I always pay attention to the system memory usage, and although it does get high during recording, nothing too alarming. It is after the recording stops where I presume buffered images during recording are then appended to the avi. Thinking that this was just a hardware error, we have tested on a variety of high-power computers. All with 128 GB DDR5 RAM, RTX 4070 - 4090, and and i7 and i9 13700k CPU. The only thing I might see to be the issue might be power supply. Most of these systems have 800 W, with the 4090 having a 1000W supply. Could that be the issue? 

Those are the two main problems we have been facing, a 200 fps cap with older spinnaker, and terrible fps processing speeds which cause recorded videos to be inaccurate with lost frames.

Any advice would be appreciated, thank you.

**Comment by bruno-f-cruz** [comment]
What settings are you using for the camera? Are you triggering the camera? We are currently using 1.29.0.5 with BFS-U3-04S2C and last time I checked, we were managing to acquire at 500hz.

**Reply by AshhadQazi** [reply]
We change the settings using the Spinview software, and last time I checked, all we have changed was just the exposure time to get maximum fps. There is an “optimize fps” button when using spinview, which just does that process for you.

**Reply by AshhadQazi** [reply]
What setting did you change for Spinnaker?

**Reply by ZhixiaoSu** [reply]
Hi, we are able to get 500f/s with harp trigger in bonsai. But there are a few parameters to take care of: 
1. In image tab, change ADC to 8 bit.
2. Lower in image tab, uncheck ISP. 
Please find other settings in the attached file images at the end. Also, another thing to check is whether it can get stable 500f/s when recording within SpinView itself. You hardware settings also seem sufficient to take care of at least two cameras, depending on the compression method. This one worked well for us: -vcodec h264_nvenc -crf 23 -preset fast -b:v 50M

![image](https://github.com/bonsai-rx/bonsai/assets/37242111/980a5607-1f39-454c-8a2d-c4376e0f0207)
![image](https://github.com/bonsai-rx/bonsai/assets/37242111/0076aed6-0b5a-4bf9-bdbc-6bcc9a62a95e)
![image](https://github.com/bonsai-rx/bonsai/assets/37242111/f63365ea-7f41-4f02-bd51-e15db4e97d77)

**Reply by AshhadQazi** [reply]
Turning Gain Auto to off and unchecking Gamma Enable seemed to fix it! The camera now records at 526 fps, and with some adjusting, I can get it to smoothen out. However, running some tests in Bonsai, there seems to be a crashing issue now? 
![image](https://github.com/bonsai-rx/bonsai/assets/170653199/6b728d4a-b0e3-4b83-a5c3-1ee9e74831b3)
It all runs smoothly and then just crashes, the only sign I can see is this odd GPU power spike right as it crashes. I think this may be an issue with Bonsai optimization, however I just switched Bonsai to be more GPU optimized, thinking that it would help.

This very well may be a compression issue as well, however I do not know how to implement what you said: "-vcodec h264_nvenc -crf 23 -preset fast -b:v 50M". Is this something I run in bonsai?

**Reply by bruno-f-cruz** [reply]
Can you include the workflow you are using to record? If it is just videowriter I find it weird that you don't crash due to out of memory after a few minutes. You might need to find a more efficient encoding strategy than the generic OpenCv one. Look for FFMPEG in these discussions for a place to start!

**Reply by AshhadQazi** [reply]
![image](https://github.com/bonsai-rx/bonsai/assets/170653199/caa3143e-e298-4a5a-9584-95691cc017eb)
This is just a very basic workflow that we have used just for testing purposes. 
![image](https://github.com/bonsai-rx/bonsai/assets/170653199/a0432475-c358-4ba7-bd02-5e1c95a7cba7)

Ill try using other videowriters, thank you!

**Reply by AshhadQazi** [reply]
Upon further testing, I switched to FFmpeg, and I am sure I have gotten it to work. However, there is still the issue of the GPU spike which just crashes Bonsai. I have ran multiple tests, and   there isn't a set point at which it crashes. So far it has only crashed after roughly 3 minutes of recording. 
![image](https://github.com/bonsai-rx/bonsai/assets/170653199/e5ea0a36-b253-4f96-a96f-0a797a295912)
This is the videowriter I am using now, FFmpeg.

**Reply by bruno-f-cruz** [reply]
Can you share the full workflow you are trying to use?

**Reply by AshhadQazi** [reply]
[Workflow.zip](https://github.com/user-attachments/files/15919467/Workflow.zip)
It is just a very simple workflow to record from spinnaker and save to a video.

**Reply by bruno-f-cruz** [reply]
The workflow you sent does not use ffmpeg

**Reply by AshhadQazi** [reply]
I don’t have access to it right now, but the ffmpeg workflow is the exact same with just the ffmpeg videowriter. But there’s the same issue with both types. Bonsai just crashes after that GPU spike

**Reply by bruno-f-cruz** [reply]
Does the spike happen if you have an empty workflow with a single `Int` operator that runs forever?

**Reply by AshhadQazi** [reply]
No it does not. It ran stable for an hour, no GPU spikes, no crashing.
